core/plugins/jdo-datanucleus-5/src/main/java/org/apache/isis/core/runtime/system/persistence/PersistenceSession5_Decouple.java

```diff
@@ -82,7 +82,13 @@ protected PersistenceSession5_Decouple(PersistenceSession5 holder, ObjectAdapter
     public ObjectAdapter adapterFor(
             final RootOid rootOid,
             final ConcurrencyChecking concurrencyChecking) {
-
+        
+        //FIXME[ISIS-1976] guard against service lookup
+        final ObjectAdapter serviceAdapter = objectAdapterContext.lookupServiceAdapterFor(rootOid);
+        if (serviceAdapter != null) {
+            return serviceAdapter;
+        }
+        
         // attempt to locate adapter for the Oid
         ObjectAdapter adapter = objectAdapterContext.lookupAdapterFor(rootOid);
         if (adapter == null) {
```

Comment:
Eventually I'm hoping that this code will simplify and then become pluggable.  Assuming that we stop supporting transient pojos, instead this code could iterate over a set of "PersistenceProviders", following the chain of responsibility pattern, where the first PersistenceProvider that recognises the format of the rootOid then takes responsibility for retrieving the corresponding pojo from its persistence store.

In the case of a PersistenceProvider for DN, that means a query to the DB.  
In the case of a PersistenceProvider for view models, it means unmarshalling the state from the oid into the pojo.  (fyi, there's also the optional SPI service, UrlEncodingService or something like, whereby the root oid is a key into some other datastore.  So really my "PersistenceProvider" is a generalization of that concept).

Reply:
added your comment inline to the source file including a FIXME